lib/statsd/instrument/aggregator.rb

```diff
@@ -0,0 +1,265 @@
+        unless thread_healthcheck
+          sink << datagram_builder(no_prefix: no_prefix).d(name, value, CONST_SAMPLE_RATE, tags)
+          return
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, type)
+
+        @mutex.synchronize do
+          if @aggregation_state.key?(key) && @aggregation_state[key].value.size + 1 >= @max_values
+            do_flush
+          end
+          if @aggregation_state.key?(key)
+            @aggregation_state[key].value << value
+          else
+            @aggregation_state[key] = AggregationValue.new(type, [value], no_prefix)
+          end
+        end
+      end
+
+      def gauge(name, value, tags: [], no_prefix: false)
+        unless thread_healthcheck
+          sink << datagram_builder(no_prefix: no_prefix).g(name, value, CONST_SAMPLE_RATE, tags)
+          return
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, GAUGE)
+
+        @mutex.synchronize do
+          @aggregation_state[key] ||= AggregationValue.new(GAUGE, value, no_prefix)
+          @aggregation_state[key].value = value
+        end
+      end
+
+      def flush
+        @mutex.synchronize { do_flush }
+      end
+
+      private
+
+      # @return Array<String>
+      EMPTY_ARRAY = [].freeze
+
+      def do_flush
+        @aggregation_state.each do |key, agg|
+          key_parts = key.split("|".b)
+          name = key_parts[0]
+          tags = key_parts[1]
+          tags = tags.empty? ? EMPTY_ARRAY : tags.split(",".b)
+          case agg.type
+          when COUNT
+            @sink << datagram_builder(no_prefix: agg.no_prefix).c(
+              name,
+              agg.value,
+              CONST_SAMPLE_RATE,
+              tags,
+            )
+          when DISTRIBUTION, MEASURE, HISTOGRAM
+            @sink << datagram_builder(no_prefix: agg.no_prefix).distribution_value_packed(
+              name,
+              agg.type.to_s,
+              agg.value,
+              CONST_SAMPLE_RATE,
+              tags,
+            )
+          when GAUGE
+            @sink << datagram_builder(no_prefix: agg.no_prefix).g(
+              name,
+              agg.value,
+              CONST_SAMPLE_RATE,
+              tags,
+            )
+          else
+            StatsD.logger.error { "[#{self.class.name}] Unknown aggregation type: #{agg.type}" }
+          end
+        end
+        @aggregation_state.clear
+      end
+
+      def tags_sorted(tags)
+        return "" if tags.nil? || tags.empty?
+
+        if tags.is_a?(Hash)
+          tags = tags.sort_by { |k, _v| k.to_s }.map! { |k, v| "#{k}:#{v}" }
+        else
+          tags.sort!
+        end
+        DatagramBuilder.normalize_tags(tags)
+      end
+
+      def packet_key(name, tags = "".b, no_prefix = false, type = COUNT)
+        "#{DatagramBuilder.normalize_string(name)}|#{tags}|#{no_prefix}|#{type}".b
+      end
```

Comment:
Create a `AggregationKey` PORO that keep these things as independent objects as ivars, and define `hash` and `eql?` so you can use it as hash key.

Reply:
Ah cool, I did not get this idea from the beginning. This sounds like a good approach.

Reply:
> Using objects for keys is a bit slower than using strings

Yeah, because the C implementation has to call back into Ruby. Once thing you could try is to make `AggregationKey` a `Struct`, this way it wouldn't need to call back in Ruby to hash the value.

Reply:
Another trick you can do, is since you know `AggregationKey` is immutable, you can precompute the hash code in `initialize` and make `hash` a simple `attr_reader`.

Reply:
Yeah, did that already. Something that also made a difference was freezing the object.